### ISSUE TITLE
Cache key function `task_input_hash` does not work with blocks

### ISSUE THREAD

**Issue by v-parkkinen** [PrefectHQ/prefect#7010]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

Using `cache_key_fn=task_input_hash` for tasks that take blocks as parameters produces different cache keys for the same block on each flow run, causing cache misses.  

### Reproduction

```python
# blocks/api_block.py
from pydantic import SecretStr
from prefect.blocks.core import Block


class ApiBlock(Block):
    base_url: str
    path: str
    api_key: SecretStr


if __name__ == "__main__":
    test_block = ApiBlock(base_url="www.example.com", path="data", api_key="secret_key")
    test_block.save("test-block")


# flow.py
from datetime import timedelta
from prefect import flow, task, get_run_logger
from prefect.tasks import task_input_hash

from blocks.api_block import ApiBlock


@task(cache_key_fn=task_input_hash, cache_expiration=timedelta(minutes=10))
def call_api_with_block(api_conf: ApiBlock):
    logger = get_run_logger()
    logger.info(f"Getting data from {api_conf.base_url}/{api_conf.path}, with key: {api_conf.api_key}")
    return 42


@flow
def get_data_from_api():
    test_block = ApiBlock.load("test-block")
    call_api_with_block(test_block)


get_data_from_api()
```

### Error

Running the above flow multiple times shows that the `call_api_with_block` task is executed every time instead of using cache.

First run:
```
08:28:14.281 | INFO    | prefect.engine - Created flow run 'sepia-beetle' for flow 'get-data-from-api'
08:28:15.559 | INFO    | Flow run 'sepia-beetle' - Created task run 'call_api_with_block-0d440d7f-0' for task 'call_api_with_block'
08:28:15.560 | INFO    | Flow run 'sepia-beetle' - Executing 'call_api_with_block-0d440d7f-0' immediately...
08:28:16.066 | INFO    | Task run 'call_api_with_block-0d440d7f-0' - Getting data from www.example.com/data, with key: **********
08:28:16.246 | INFO    | Task run 'call_api_with_block-0d440d7f-0' - Finished in state Completed()
08:28:16.419 | INFO    | Flow run 'sepia-beetle' - Finished in state Completed('All states completed.')
```
Second run:
```
08:28:37.289 | INFO    | prefect.engine - Created flow run 'dark-lion' for flow 'get-data-from-api'
08:28:38.557 | INFO    | Flow run 'dark-lion' - Created task run 'call_api_with_block-0d440d7f-0' for task 'call_api_with_block'
08:28:38.558 | INFO    | Flow run 'dark-lion' - Executing 'call_api_with_block-0d440d7f-0' immediately...
08:28:39.061 | INFO    | Task run 'call_api_with_block-0d440d7f-0' - Getting data from www.example.com/data, with key: **********
08:28:39.236 | INFO    | Task run 'call_api_with_block-0d440d7f-0' - Finished in state Completed()
08:28:39.417 | INFO    | Flow run 'dark-lion' - Finished in state Completed('All states completed.')
```

### Versions

```
Version:             2.4.2
API version:         0.8.0
Python version:      3.9.13
Git commit:          65807e84
Built:               Fri, Sep 23, 2022 10:43 AM
OS/Arch:             darwin/arm64
Profile:             nmt
Server type:         cloud
```

### Additional context

Running the same task but using strings as input makes caching work as expected:
```
@task(cache_key_fn=task_input_hash, cache_expiration=timedelta(minutes=10))
def call_api_with_strings(base_url: str, path: str, api_key: SecretStr):
    logger = get_run_logger()
    logger.info(f"Getting data from {base_url}/{path}, with key: {api_key} (call_api_with_strings)")
    return 42
    
    
@flow
def get_data_from_api():
    test_block = ApiBlock.load("test-block")
    call_api_with_strings(test_block.base_url, test_block.path, test_block.api_key)    
```

Wrapping `task_input_hash` with a custom function and logging the cache key shows that the key produced by `task_input_hash` is different on each flow run:
```
def cache_key_logger(context, params):
    logger = get_run_logger()
    cache_key = task_input_hash(context, params)
    logger.info(cache_key)
    return cache_key
    
@task(cache_key_fn=cache_key_logger, cache_expiration=timedelta(minutes=10))
def call_api_with_block(api_conf: ApiBlock):
    ...
```

**Comment by v-parkkinen** [comment]
On closer inspection it seems that this issue is caused by the `__fields_set__` property of Pydantic models. From [Pydantic docs](https://pydantic-docs.helpmanual.io/usage/models/#model-properties):
```
__fields_set__
Set of names of fields which were set when the model instance was initialised
```
The order of elements in this set changes between flow runs, causing the block to be pickled (and consequently hashed) differently on each run.

**Comment by madkinsz** [comment]
Thanks for the issue and investigation! I think we should solve this by adding explicit support for Pydantic models to `task_inputs_hash` (or one of our downstream caching utilities) rather than using pickle. Contributions are welcome here :)

**Comment by sharmadhiraj86** [comment]
Hii @madkins, I am interested in working on this issue. Can you please assign this issue to me?

**Comment by madkinsz** [comment]
Hi @sharmadhiraj86 — Thanks for checking in, but we've already got a fix merged in #7048.
